freud/cluster.pyx

```diff
@@ -70,15 +73,20 @@ cdef class Cluster(Compute):
             A list of lists of the keys contained in each cluster.
     """
     cdef freud._cluster.Cluster * thisptr
-    cdef r_max
+    cdef float r_max
+    cdef public freud.util.ManagedArrayManager __cluster_idx
 
     def __cinit__(self, float r_max):
         self.thisptr = new freud._cluster.Cluster(r_max)
         self.r_max = r_max
+        self.__cluster_idx = freud.util.ManagedArrayManager.init(
```

Comment:
This property `__cluster_idx` and its management feels super hacky. I'm not sure if the Cython side is as clean as it could be, yet. I'd like to think about this some more before I can make a recommendation for improvement. A few possible strategies:
1. Simplify more of the logic by leveraging the base `Compute` class?
2. Reduce the number of times we have to specify array dtypes?
3. Wrap more logic into `util`?

Reply:
I'm open to any ideas on how to simplify this part of the code, I agree that I"m not a huge fan of how the Cython side of things is currently structured. Suggestions 1 and 3 are kind of vague though, I'm not really sure what you're proposing. Regarding 2, I don't think dtypes are specified in very many places... unless you're referring to template types? There's not much we can do about needing to specify the template parameters for ManagedArray, but I've tried to encapsulate all of that into util.pxd.

Reply:
Suggestions for improvement:

1. Use single-underscores instead of double-underscores for the private variable.
2. Create an array "registration" method, inherited from `Compute`, and use it in the `__cinit__`.
```python
self.register_array(self._cluster_idx, self.thisptr.getClusterIdx())
```
3. Make the array know its intended shape on the C++ side, perhaps in a member variable like `vector<size_t> m_shape`, and use this in conjunction with the `reshape` method I proposed below. This could be integrated with Index2D and Index3D so they can return a vector, like `m_array.set_shape(index.get_shape())`.